### DISCUSSION TITLE
Compile a "freestanding" WASM binary

### DISCUSSION THREAD

**Discussion by xylobol** [dotnet/runtime#102427] [Q&A] [answered]
So, here's a fun one: I'm looking to build a WASM binary, but one that doesn't use WASI or any JS interop. I'm doing some weird low-level stuff with this - the details aren't important - but I'm looking to do something like publish a class library that exports an `add(i32, i32)` function, which can then be called from the WASM host. Doing it the other way around would also be nice - importing functions/modules which have already been linked in through something like `LibraryImport`. I'm expecting to lose most (all?) of the standard library by doing this.

**Comment by hez2010** [comment]
You need to use NativeAOT-LLVM in the lab repo: https://github.com/dotnet/runtimelab/tree/feature/NativeAOT-LLVM/samples/NativeLibrary

**Comment by maraf** [comment]
This is what WASI (or `wasi-wasm` RID in .NET) is for. With WASI preview 2 and components you will get a more standardized solution for this. Neither Mono (wasi-experimental wokload developed here) nor NativeAOT-LLVM (community driven project) does support preview 2. With NativeAOT-LLVM you can accomplish your case, with .NET 9 preview SDK you should be able to it with Mono as well. 

In both cases linker will be run removing unused bits of standard library.

cc @SingleAccretion

**Reply by SingleAccretion** [reply, marked as answer]
I will add that if you want something truly "freestanding", i. e. no WASI imports in the final binary - you mention "one that doesn't use WASI", this would effectively require creating your own flavor of .NET runtime. It is possible, e. g. https://github.com/MichalStrehovsky/zerosharp, but you will be largely on your own.
